specs/data-grid/data-grid.spec.md

```diff
@@ -0,0 +1,302 @@
+- `generateColumns(object): DataGridColumn`   
+Static function that creates a basic set of columns from an object representing a row.  
+
+*Events*
+- none
+
+*parts:*
+- `headerSlot`
+
+- `rowsSlot`
+
+
+**Data grid header**
+- `fast-data-grid-header`
+
+*Attributes:*
+- none
+
+*properties:*
+- `columnsData`  
+An array of `DataGridColumn` objects that define what columns will be displayed in the grid.  The order of the columns determines their order in the grid.
+
+- `headerCellItemTemplate`  
+Custom [template](https://fast.design/docs/fast-element/declaring-templates) to use when generating header cells by iterating over data.  The default template uses `fast-data-grid-header-cell`, this is where authors change that.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `headerCellsSlot`
+
+
+**Data grid header cell**
+- `fast-data-grid-header-cell`
+
+*Attributes:*
+- extends HTML Element attributes
+
+*properties:*
+- `columnData`  
+The `DataGridColumn` this column header represents.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `headerCellsSlot`
+
+
+**Data grid row**
+- `fast-data-grid-row`
+
+*Attributes:*
+- `gridTemplateColumns`  
+String that gets applied to the the css gridTemplateColumns attribute for the row
+
+*properties:*
+- `rowData`  
+The objects that contains the data to be displayed in this row.
+
+- `columnsData`  
+An array of `DataGridColumn` objects that define what columns will be displayed in the grid.  The order of the columns determines their order in the grid.
+
+- `cellItemTemplate`  
+Custom [template](https://fast.design/docs/fast-element/declaring-templates) to use when generating cells by iterating over data.  The default template uses `fast-data-grid-cell`, this is where authors change that.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `cellsSlot`
+
+**Data grid cell**
+*Component name:*
+- `fast-data-grid-cell`
+
+*Attributes:*
+- `gridColumnIndex`  
+The grid column this cell appears in.
+
+*properties:*
+- `rowData`  
+The object that contains the data to be displayed in this row.
+
+- `columnData`  
+The `DataGridColumn` this cell represents.
+
+*Slots:*
+- `default`  
+Default slot for items
+
+*Events*
+- none
+
+*parts:*
+- `cellSlot`
+
+## Implementation
+
+The most basic use case for this component would be to just pass it an array of objects for it to render.  
+
+For example given a grid component:
+```html
+ <fast-data-grid id="defaultGrid"></fast-data-grid>
+ ```
+
+ And some data:
+ ```js
+const baseRows: object[] = [
+    { name: "Rob", age: "19" },
+    { name: "Bob", age: "20" },
+];
+ ```
+
+ An author could pass the data to the component from a javascript function:
+
+```js
+onLoad(): void {
+   const defaultGrid: DataGrid | null = document.getElementById(
+        "defaultGrid"
+    ) as DataGrid;
+    if (defaultGrid !== null) {
+        defaultGrid.rowsData = baseRows;
```

Comment:
Is this the first component that has had to have a property set to work, or have we done this in other places and is it an expected pattern? I'm not arguing against it since I think the pattern makes sense here, just curious as to our current component line up and how this will have to affect tooling since we currently assume all components can be implemented sans additional JavaScript implementation.

Reply:
I initialially had attributes that took data in the form of JSON strings but got feedback from @EisenbergEffect that "web compoent guidance recommends not creating attrs that accept json but using properties that take objects instead".  I kinda prefer the data only approach myself.

Reply:
Agreed.